packages/autoloader/src/class-container.php

```diff
@@ -0,0 +1,139 @@
+<?php
+/* HEADER */ // phpcs:ignore
+
+/**
+ * This class manages the files and dependencies of the autoloader.
+ */
+class Container {
+
+	/**
+	 * Since each autoloader's class files exist within their own namespace we need a map to
+	 * convert between the local class and a shared key. Note that no version checking is
+	 * performed on these dependencies and the first autoloader to register will be the
+	 * one that is utilized.
+	 */
+	const SHARED_DEPENDENCY_KEYS = array(
+		Hook_Manager::class => 'Hook_Manager',
+	);
+
+	/**
+	 * A map of all the dependencies we've registered with the container and created.
+	 *
+	 * @var array
+	 */
+	protected $dependencies;
+
+	/**
+	 * The constructor.
+	 */
+	public function __construct() {
+		$this->dependencies = array();
+
+		$this->register_shared_dependencies();
+		$this->register_dependencies();
+		$this->initialize_globals();
+	}
+
+	/**
+	 * Gets a dependency out of the container.
+	 *
+	 * @param string $class The class to fetch.
+	 *
+	 * @return mixed
+	 * @throws \InvalidArgumentException When a class that isn't registered with the container is fetched.
+	 */
+	public function get( $class ) {
+		if ( ! isset( $this->dependencies[ $class ] ) ) {
+			throw new \InvalidArgumentException( "Class '$class' is not registered with the container." );
+		}
+
+		return $this->dependencies[ $class ];
+	}
+
+	/**
+	 * Registers all of the dependencies that are shared between all instances of the autoloader.
+	 */
+	private function register_shared_dependencies() {
+		global $jetpack_autoloader_container_shared;
+		if ( ! isset( $jetpack_autoloader_container_shared ) ) {
+			$jetpack_autoloader_container_shared = array();
+		}
+
+		$key = self::SHARED_DEPENDENCY_KEYS[ Hook_Manager::class ];
+		if ( ! isset( $jetpack_autoloader_container_shared[ $key ] ) ) {
+			require_once __DIR__ . '/class-hook-manager.php';
+			$jetpack_autoloader_container_shared[ $key ] = new Hook_Manager();
```

Comment:
I thought about the shared `Hook_Manager` object a lot. I'm not enthusiastic about the fact that we could use an older version of the `Hook_Manager` class if an older version of the autoloader is registered first. However, I think that would just be a temporary situation, and it would be resolved on the next page load, right?

This situation would only happen when the plugin with the latest autoloader was not in the active plugins list (for example, it's an mu-plugin or it's activating in a way that couldn't be detected). The plugin would be added to the cache file. Then its autoloader would be identified as the latest autoloader in subsequent page loads, and the latest version of the `Hook_Manager` class would be used. Is that correct?

Reply:
> I'm not enthusiastic about the fact that we could use an older version of the Hook_Manager class if an older version of the autoloader is registered first.

My rationale here was that it seemed incredibly unlikely we would ever change the `Hook_Manager` class. I'd considered some form of version reconciliation but given how "complete" the autoloader feels I figured we should cross that bridge if we get there.

> Is that correct?

Yep! The only caveat is that the newer autoloader will need to be aware of that limitation and act accordingly.

**Proposed Solution**

The solution to this problem I'd entertained was an interface for all shared dependencies that declared a `getProperties()` method and a `VERSION` constant string defined as part of the autoloader generation. The idea is that the constructor of newer classes can consume the older class and receive its properties in order to replace it.

If you think this is something we should do now though I'm not opposed to adding it.

Reply:
I don't think it's necessary to do that now. As you pointed out, it seems unlikely that `Hook_Manager` will need to be changed.